### PR TITLE
Allow management-cluster creation using ClusterClass when `package-based-lcm` flag is enabled

### LINKED ISSUE #2358: Support creating management-cluster using package-based-lcm and with cluster-class
**Describe the feature request**

Request to add support for creating cluster-class based management-cluster when `package-based-lcm` featureflag is enabled.

Currently, we have added support for creation of workload cluster using clusterclass however the managment-cluster gets created in legacy way. We need to add the support for creating management-cluster using cluster-class.

**Describe alternatives you've considered**
none

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [x] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it

- This PR fixes issues for creating management-cluster using `package-based-lcm` approach
- Fixes includes:
    - Set featuregate for addons ClusterBootstrapController to true
    - Install clusterclass to tkg-system namespace
    - Fix kapp-controller package values
    - Install kapp-controller before installing providers
    - Do not wait for addons deployment if `package-based-lcm` feature flag is enabled
    - Apply overlays based on pblcm flag and Remove unneccesary overlays from yttcc directory
    - Add patch awsclustertemplate for cni-ingress as part of clusterclass
    - Create cluster template after deploying clusterclass with packages
    - Do not wait for packages on mgmt-cluster if package-based-lcm ff is enabled

…

### Which issue(s) this PR fixes

Fixes #2358

### Describe testing done for PR

1. Build all management-components along with management package repository
2. Use the management package repository to create management-cluster
3. Make sure management-cluster gets created successfully using clusterclass

### Release note

```release-note
package-based-lcm: Allow management-cluster creation using ClusterClass
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/tkg/managementcomponents/management_component_install.go [hunk 1]
```diff
@@ -55,8 +55,14 @@ func InstallManagementComponents(mcip *ManagementComponentsInstallOptions) error
 	if err != nil {
 		return errors.Wrap(err, "unable to get cluster client")
 	}
-	if err := InstallKappController(clusterClient, mcip.KappControllerOptions); err != nil {
-		return errors.Wrap(err, "unable to install kapp-controller")
+
+	resouceFilePreDeploy := os.Getenv("_ADDITIONAL_MANAGEMENT_COMPONENT_CONFIGURATION_FILE_PRE_DEPLOY")
```
Comment (anujc25) [inline pkg/v1/tkg/managementcomponents/management_component_install.go:59]: Add comment that this is temporary and will be removed in future.

File: pkg/v1/tkg/managementcomponents/management_component_install.go [hunk 2]
```diff
@@ -75,6 +81,15 @@ func InstallManagementComponents(mcip *ManagementComponentsInstallOptions) error
 		return errors.Wrap(err, "timed out waiting for management packages to get reconciled successfully")
 	}
 
+	resouceFilePostDeploy := os.Getenv("_ADDITIONAL_MANAGEMENT_COMPONENT_CONFIGURATION_FILE_POST_DEPLOY")
```
Comment (anujc25) [inline pkg/v1/tkg/managementcomponents/management_component_install.go:84]: Add comment that this is temporary and will be removed in future.

**Comment by vuil** [comment]
thanks, changes looks good in general. My first gripe is that it is quite a big change. I wonder if we can tease them apart some. 
nit: it would be nice if we have at most one commit to sync the providers, but I can see the sequence of changes makes it hard.

more importantly: can the big change to remove lots of addons-related stuff be independently posted, reviewed and checked in first. They are making it hard to review the effects other overlay changes. (in https://storage.googleapis.com/tkg-clustergen/2286/20220512215847/clustergen.diff.txt)  also introduced in this PR

**Comment by anujc25** [comment]
Created Separate PR for removing additional overlays from `yttcc`. https://github.com/vmware-tanzu/tanzu-framework/pull/2379

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 3]
```diff
@@ -93,6 +93,20 @@ spec:
         default: false
   - name: network
     required: true
+    schema:
+      openAPIV3Schema:
+        type: string
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:98]: This is not right. 
```
- name: network
  required: true
  ```
  should be where the new `- name: proxy` is (at line 109) and the schema at line 96-98 should be removed

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 4]
```diff
@@ -279,6 +293,57 @@ spec:
             {{- end }}
             {{- end }}
 
+  - name: AWSCT_CNIIngressRules_Calico
+    enabledIf: '{{if (eq .cni "calico") }}true{{end}}'
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:297]:
```suggestion
    enabledIf: '{{ .cni | eq "calico" }}'
```

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 5]
```diff
@@ -279,6 +293,57 @@ spec:
...
+        valueFrom:
+          template: |
+            description: bgp (calico)
+            fromPort: 179
+            protocol: tcp
+            toPort: 179
+      - op: add
+        path: "/spec/template/spec/network/cni/cniIngressRules/-"
+        valueFrom:
+          template: |
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:316]: This doesn't actually have a templated value right? couldn't this be
```
value: |
  description: IP-in-IP (calico)
  ...
```

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 6]
```diff
@@ -279,6 +293,57 @@ spec:
...
+    - selector:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: AWSClusterTemplate
+        matchResources:
+          infrastructureCluster: true
+      jsonPatches:
+      - op: add
+        path: "/spec/template/spec/network/cni/cniIngressRules/-"
+        valueFrom:
+          template: |
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:333]: Same question about templated value, can't this just be a value directly/

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 7]
```diff
@@ -279,6 +293,57 @@ spec:
...
+            description: antrea1
+            fromPort: 10349
+            protocol: tcp
+            toPort: 10349
+      - op: add
+        path: "/spec/template/spec/network/cni/cniIngressRules/-"
+        valueFrom:
+          template: |
+            description: genev
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:342]: Same question about templated value

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 8]
```diff
@@ -279,6 +293,57 @@ spec:
...
+      - op: add
+        path: "/spec/template/spec/network/cni/cniIngressRules/-"
+        valueFrom:
+          template: |
+            description: IP-in-IP (calico)
+            fromPort: -1
+            protocol: "4"
+            toPort: 65535
+  - name: AWSCT_CNIIngressRules_Antrea
+    enabledIf: '{{if (eq .cni "antrea") }}true{{end}}'
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:322]:
```suggestion
    enabledIf: '{{ .cni | eq "antrea" }}'
```

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml [hunk 9]
```diff
@@ -279,6 +293,57 @@ spec:
...
+    enabledIf: '{{if (eq .cni "calico") }}true{{end}}'
+    definitions:
+    - selector:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: AWSClusterTemplate
+        matchResources:
+          infrastructureCluster: true
+      jsonPatches:
+      - op: add
+        path: "/spec/template/spec/network/cni/cniIngressRules/-"
```
Comment (tenczar) [inline pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml:306]: My style guide recommends not using `"` when not necessary, this should just be `path: /spec/template/spec/network/cni/cniIngressRules/-`

**Review by tenczar — CHANGES_REQUESTED** [review]
Looks good. Just a couple small comments from me.

**Comment by vuil** [comment]
This change warrants some documentation (maybe in docs/cli/tkgctl) regarding the changes in the MC deployment approach. Please outline the sequence of various how various components are installed in the bootstrap and management cluster. Also mention the hack re: _ADDITIONAL_MANAGEMENT_COMPONENT_CONFIGURATION_FILE (to removed when the implementation no longer needs it)

**Review by tenczar — APPROVED** [review]
LGTM

**Comment by vuil** [comment]
Please review https://storage.googleapis.com/tkg-clustergen/2286/20220516181138/clustergen.diff.txt,
There are a bunch of testcases that turned from positive to negative:
e.g.
diff -r -U15 old/06364262.case.cc.output new/06364262.case.cc.output
PTAL to see if that is intendedwith the change (usually not)

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml (hunk 5 above)
Comment (anujc25) [reply pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml]: Looks like setting this to value doesn't work.
```
  - spec.patches[1].definitions[0].jsonPatches[0].valueFrom: Invalid value: "{}": valueFrom must set either template or variable
```

**Review by tenczar — APPROVED** [review]

File: pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml (hunk 5 above)
Comment (anujc25) [reply pkg/v1/providers/infrastructure-aws/v1.2.0/cconly/base.yaml]: Needed to do remove `|` when using `value` and not use `valueFrom`. Updated

**Review by vuil — APPROVED** [review]
lgtm, thanks!
Good that we root-caused the discreptancy shown in the clustergen diffs

**Pull request merged**